### PR TITLE
feat(demo): Demo visualizer for buffered ranges.

### PR DESCRIPTION
**Pull request by theodab**
You can activate this visualizer UI by using the new
"Buffer Visualizer" button in the overflow menu of the player UI.

### CODE REVIEW

File: demo/visualizer.js [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    const ctx = /** @type {CanvasRenderingContext2D} */ (
+      this.canvas_.getContext('2d'));
+    ctx.imageSmoothingEnabled = false;
+
+    // Make a black background.
+    ctx.fillStyle = '#000000';
+    ctx.fillRect(0, 0, this.canvas_.width, this.canvas_.height);
+
+    // Define the colors.
+    const colors = ['#FF0000', '#00FF00', '#0000FF'];
```
Comment (avelad) [inline demo/visualizer.js:273]: I need a comment to know what each color means

File: lib/player.js [hunk 2]
```diff
@@ -5500,14 +5502,22 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
...
     }
     this.pollBufferState_();
+
+    // Dispatch an event for users to consume, too.
+    const data = new Map()
+        .set('start', start)
+        .set('end', end)
+        .set('contentType', contentType);
+    this.dispatchEvent(this.makeEvent_(
+        shaka.util.FakeEvent.EventName.SegmentAppended, data));
```
Comment (avelad) [inline lib/player.js:5520]: This event needs to be documented at the beginning of the player.js

File: demo/visualizer.js [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+    // Listen for when new buffers are appended.
+    player.addEventListener('segmentAppended', (event) => {
+      const start = /** @type {number} */ (event['start']);
+      const end = /** @type {number} */ (event['end']);
+      const contentType = /** @type {string} */ (event['contentType']);
+      let isAudio = false;
+      if (contentType == 'audio') {
+        isAudio = true;
+      } else if (contentType != 'video') {
+        // If it's not video or audio, either it's text or it's invalid.
```
Comment (avelad) [inline demo/visualizer.js:81]: why not display the text? If the text is segmented it may be useful to have this information.

**Review by avelad — CHANGES_REQUESTED** [review]

**Comment by avelad** [comment]
Note: I have tested with HLS in Safari and it shows video, not video and audio.

**Comment by theodab** [comment]
> Note: I have tested with HLS in Safari and it shows video, not video and audio.

That's odd, to see one but not the other. You mean with `streaming.useNativeHlsOnSafari` disabled, right?

**Comment by avelad** [comment]
> > Note: I have tested with HLS in Safari and it shows video, not video and audio.
> 
> That's odd, to see one but not the other. You mean with `streaming.useNativeHlsOnSafari` disabled, right?

`streaming.useNativeHlsOnSafari = true`

**Comment by theodab** [comment]
> > > Note: I have tested with HLS in Safari and it shows video, not video and audio.
> > 
> > 
> > That's odd, to see one but not the other. You mean with `streaming.useNativeHlsOnSafari` disabled, right?
> 
> `streaming.useNativeHlsOnSafari = true`

Well, the segment appended updates (the pieces of text that show up alongside the colored bars) rely on activity within `StreamingEngine`, so that feature not working properly with native HLS is expected. If anything I'm surprised that it shows video.

**Review by avelad — APPROVED** [review]

**Pull request merged**